// @xmtp/xmtp-js/dist/cjs/src/crypto/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletSigner = exports.Signature = exports.PrivateKeyBundleV2 = exports.PrivateKeyBundleV1 = exports.decodePrivateKeyBundle = exports.SignedPrivateKey = exports.PrivateKey = exports.PublicKeyBundle = exports.PublicKey = exports.SignedPublicKeyBundle = exports.SignedPublicKey = exports.UnsignedPublicKey = exports.decrypt = exports.encrypt = exports.utils = void 0;
const PublicKeyBundle_1 = require("./PublicKeyBundle");
Object.defineProperty(exports, "PublicKeyBundle", { enumerable: true, get: function () { return PublicKeyBundle_1.PublicKeyBundle; } });
Object.defineProperty(exports, "SignedPublicKeyBundle", { enumerable: true, get: function () { return PublicKeyBundle_1.SignedPublicKeyBundle; } });
const PrivateKey_1 = require("./PrivateKey");
Object.defineProperty(exports, "SignedPrivateKey", { enumerable: true, get: function () { return PrivateKey_1.SignedPrivateKey; } });
Object.defineProperty(exports, "PrivateKey", { enumerable: true, get: function () { return PrivateKey_1.PrivateKey; } });
const PrivateKeyBundle_1 = require("./PrivateKeyBundle");
Object.defineProperty(exports, "PrivateKeyBundleV1", { enumerable: true, get: function () { return PrivateKeyBundle_1.PrivateKeyBundleV1; } });
Object.defineProperty(exports, "PrivateKeyBundleV2", { enumerable: true, get: function () { return PrivateKeyBundle_1.PrivateKeyBundleV2; } });
Object.defineProperty(exports, "decodePrivateKeyBundle", { enumerable: true, get: function () { return PrivateKeyBundle_1.decodePrivateKeyBundle; } });
const PublicKey_1 = require("./PublicKey");
Object.defineProperty(exports, "UnsignedPublicKey", { enumerable: true, get: function () { return PublicKey_1.UnsignedPublicKey; } });
Object.defineProperty(exports, "SignedPublicKey", { enumerable: true, get: function () { return PublicKey_1.SignedPublicKey; } });
Object.defineProperty(exports, "PublicKey", { enumerable: true, get: function () { return PublicKey_1.PublicKey; } });
const Signature_1 = __importStar(require("./Signature"));
exports.Signature = Signature_1.default;
Object.defineProperty(exports, "WalletSigner", { enumerable: true, get: function () { return Signature_1.WalletSigner; } });
const utils = __importStar(require("./utils"));
exports.utils = utils;
const encryption_1 = require("./encryption");
Object.defineProperty(exports, "encrypt", { enumerable: true, get: function () { return encryption_1.encrypt; } });
Object.defineProperty(exports, "decrypt", { enumerable: true, get: function () { return encryption_1.decrypt; } });
