// @xmtp/proto/ts/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.invitation = exports.contact = exports.ciphertext = exports.signature = exports.publicKey = exports.privateKey = exports.composite = exports.xmtpEnvelope = exports.authn = exports.fetcher = exports.messageApi = void 0;
exports.messageApi = __importStar(require("./message_api/v1/message_api.pb"));
exports.fetcher = __importStar(require("./fetch.pb"));
exports.authn = __importStar(require("./message_api/v1/authn.pb"));
exports.xmtpEnvelope = __importStar(require("./message_contents/xmtp_envelope.pb"));
exports.composite = __importStar(require("./message_contents/composite.pb"));
exports.privateKey = __importStar(require("./message_contents/private_key.pb"));
exports.publicKey = __importStar(require("./message_contents/public_key.pb"));
exports.signature = __importStar(require("./message_contents/signature.pb"));
exports.ciphertext = __importStar(require("./message_contents/ciphertext.pb"));
exports.contact = __importStar(require("./message_contents/contact.pb"));
exports.invitation = __importStar(require("./message_contents/invitation.pb"));
