// @xmtp/xmtp-js/dist/cjs/src/codecs/Text.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextCodec = exports.Encoding = exports.ContentTypeText = void 0;
const MessageContent_1 = require("../MessageContent");
// xmtp.org/text
//
// This content type is used for a plain text content represented by a simple string
exports.ContentTypeText = new MessageContent_1.ContentTypeId({
    authorityId: 'xmtp.org',
    typeId: 'text',
    versionMajor: 1,
    versionMinor: 0,
});
var Encoding;
(function (Encoding) {
    Encoding["utf8"] = "UTF-8";
})(Encoding = exports.Encoding || (exports.Encoding = {}));
class TextCodec {
    get contentType() {
        return exports.ContentTypeText;
    }
    encode(content) {
        return {
            type: exports.ContentTypeText,
            parameters: { encoding: Encoding.utf8 },
            content: new TextEncoder().encode(content),
        };
    }
    decode(content) {
        const encoding = content.parameters.encoding;
        if (encoding && encoding !== Encoding.utf8) {
            throw new Error(`unrecognized encoding ${encoding}`);
        }
        return new TextDecoder().decode(content.content);
    }
}
exports.TextCodec = TextCodec;
